sound/soc/sof/ipc.c

```diff
@@ -340,6 +340,56 @@ static int sof_ipc_tx_message_unlocked(struct snd_sof_ipc *ipc,
 	return tx_wait_done(ipc, msg, reply_data);
 }
 
+/**
+ * sof_ipc_send_msg - generic function to prepare and send one IPC message
+ * @sdev:		pointer to SOF core device struct
+ * @msg_data:		pointer to a message to send
+ * @msg_bytes:		number of bytes in the message
+ * @reply_bytes:	number of bytes available for the reply
```

Comment:
@plbossart, neither ;)
This is a helper function to set up the `struct snd_sof_ipc_msg` at `sdev->ipc->msg` with the needed parameters for sending a message and then also make a call to the platform code make the transfer.
`msg_data` and `msg_bytes` is the message we send and it is passed via the `msg`, for the reply we only tell the platform code the maximum available space given by the caller to receive back the reply, the pointer to the reply is not given as the reply content handover is in the IPC specific code (`sof_ipc3_get_reply`, AMD have this open-coded).

The comment is actually correct, it is the bytes available for the reply.

Reply:
not following the explanation, sorry. the caller provides the number of bytes for the reply, but not the pointer for the reply?

Reply:
> not following the explanation, sorry. the caller provides the number of bytes for the reply, but not the pointer for the reply?

The `sof_ipc_send_msg()` is a helper function to fill the `sdev->ipc->msg` struct as it is a generic struct. We could have this as a big #define or strict inline function if we want, or just duplicate it for every IPC version.

The caller is the one who calls the `sof_ipc_tx_message()`, there we have the reply pointer and the size allocated for it. This caller provided reply data pointer is only used when we got the reply into the sdev->ipc->msg.reply_data (which is allocated at probe time) and it is processed in the IPC code and finally we copy from the intermediate storage to the user provided buffer.

The message sending itself does not need the caller provided reply buffer, the sending itself does not even needs to know the size the caller provided.
The size is needed when we got the interrupt that the message has been processed, and we can read out the reply to the temporary buffer. Here we check the caller's provided size for the reply and do minimal copy and size checks.
As a last step the we copy the data to the caller's buffer, but it is done via the completion handling in the tx path (we wait for the signal that we got the reply).

It is complicated and simple at the same time, but I really don't know how else it could be handled without major changes in the logic.

Reply:
thanks for the explanations @ujfalusi. I never realized that we had an intermediate msg->reply_data intermediate buffer and a memcpy.